src/utils.zig

```diff
@@ -31,35 +33,62 @@ pub const toMagicNumberLittle = switch (native_endian) {
+    const T = @typeInfo(@TypeOf(data)).Pointer.child;
+    inline for (meta.fields(T)) |entry| {
+        switch (@typeInfo(entry.field_type)) {
+            .Enum => {
+                const value = @enumToInt(@field(data, entry.name));
+                _ = try std.meta.intToEnum(entry.field_type, value);
+            },
+            .Struct => {
+                try checkEnumFields(&@field(data, entry.name));
+            },
+            else => {},
+        }
+    }
 }
 
-pub fn readStructForeign(reader: io.StreamSource.Reader, comptime T: type) !T {
-    comptime std.debug.assert(@typeInfo(T).Struct.layout != builtin.TypeInfo.ContainerLayout.Auto);
-
-    var result: T = undefined;
+pub fn readStructNative(reader: io.StreamSource.Reader, comptime T: type) StructReadError!T {
+    var result: T = try reader.readStruct(T);
+    try checkEnumFields(&result);
+    return result;
+}
 
+fn swapFieldBytes(data: anytype) error{InvalidEnumTag}!void {
+    const T = @typeInfo(@TypeOf(data)).Pointer.child;
     inline for (meta.fields(T)) |entry| {
         switch (@typeInfo(entry.field_type)) {
-            .ComptimeInt, .Int => {
-                @field(result, entry.name) = try reader.readIntForeign(entry.field_type);
+            .Int => |int| {
+                if (int.bits > 8) @field(data, entry.name) = @byteSwap(entry.field_type, @field(data, entry.name));
             },
             .Struct => {
-                @field(result, entry.name) = try readStructForeign(reader, entry.field_type);
+                try swapFieldBytes(&@field(data, entry.name));
             },
             .Enum => {
-                @field(result, entry.name) = try reader.readEnum(entry.field_type, switch (native_endian) {
-                    builtin.Endian.Little => builtin.Endian.Big,
-                    builtin.Endian.Big => builtin.Endian.Little,
-                });
+                const value = @enumToInt(@field(data, entry.name));
+                if (@bitSizeOf(@TypeOf(value)) > 8) {
+                    @field(data, entry.name) = try std.meta.intToEnum(entry.field_type, @byteSwap(@TypeOf(value), value));
+                } else {
+                    _ = try std.meta.intToEnum(entry.field_type, value);
+                }
+            },
+            .Array => |array| {
+                if (array.child != u8) {
+                    @compileError("Add support for type " ++ @typeName(T) ++ "." ++ @typeName(entry.field_type) ++ " in swapFieldBytes");
+                }
             },
+            .Bool => {},
             else => {
-                @compileError(std.fmt.comptimePrint("Add support for type {} in readStructForeign", .{@typeName(entry.field_type)}));
+                @compileError("Add support for type " ++ @typeName(T) ++ "." ++ @typeName(entry.field_type) ++ " in swapFieldBytes");
             },
         }
     }
+}
 
+pub fn readStructForeign(reader: io.StreamSource.Reader, comptime T: type) StructReadError!T {
```

Comment:
I really like this approach :) I wonder if it will be more performant doing it this way.

Reply:
If we had a buffered reader it probably wouldn't make much a difference but as it is now it might. Although I read that syscalls are for some reason much cheaper on Linux than on Windows (or at least some of them)